src/c_kzg_4844.c

```diff
@@ -1098,34 +1100,46 @@ static C_KZG_RET compute_challenges(BLSFieldElement *out, BLSFieldElement r_powe
   const size_t np = ni + n * BYTES_PER_BLOB;
   const size_t nb = np + n * 48;
 
-  uint8_t* bytes = calloc(nb + (n == 0), sizeof(uint8_t)); // need at least 1 byte more than ni for hash later
+  uint8_t* bytes = calloc(nb, sizeof(uint8_t));
   if (bytes == NULL) return C_KZG_MALLOC;
 
+  /* Copy domain seperator */
   memcpy(bytes, FIAT_SHAMIR_PROTOCOL_DOMAIN, 16);
   bytes_of_uint64(&bytes[16], FIELD_ELEMENTS_PER_BLOB);
   bytes_of_uint64(&bytes[16 + 8], n);
 
+  /* Copy polynomials */
   for (i = 0; i < n; i++)
     for (j = 0; j < FIELD_ELEMENTS_PER_BLOB; j++)
       bytes_from_bls_field(&bytes[ni + BYTES_PER_FIELD_ELEMENT * (i * FIELD_ELEMENTS_PER_BLOB + j)], &polys[i][j]);
 
+  /* Copy commitments */
   for (i = 0; i < n; i++)
     bytes_from_g1(&bytes[np + i * 48], &comms[i]);
 
-  uint8_t hash_output[32] = {0};
-  hash(hash_output, bytes, nb);
-  memcpy(bytes, hash_output, 32);
-  bytes[32] = 0x0;
-  hash(hash_output, bytes, 33);
-
-  if (n > 0) {
-    if (n > 1) bytes_to_bls_field(&r_powers[1], hash_output);
-    compute_powers(r_powers, n);
-  }
-
-  bytes[32] = 0x1;
-  hash(hash_output, bytes, 33);
-  bytes_to_bls_field(out, hash_output);
+  /* Now let's create challenges! */
+  uint8_t hashed_data[32] = {0};
+  hash(hashed_data, bytes, nb);
+
+  /* Compute r */
+  uint8_t r_bytes[32] = {0};
+  uint8_t hash_input_0[33] = {0}; // hashed_data + b'\x00'
```

Comment:
This one I would prefer to reuse (just one `hash_input` instead of two). I'm fine with not re-using `bytes` to improve readability though.

Reply:
For the record, stack allocation is an extremely cheap (like [usually a single CPU instruction](https://stackoverflow.com/a/161073)) so IMO this is mostly a question of readability than performance. If you think the code is cleaner with a single `hash_input` then let's go with that. Other than that, the performance overhead should be negligible and we have the luxury to optimize for niceness :)

(To be frank, in the above there is also the extra tiny overhead of the zeroing with `= {0};`. In this code the zeroing is not really necessary as we are writing over the entire array, and can be omitted for extra (negligible) "speed". Negligible because memset is also extremely fast.)

Reply:
Done in bdfa79c